src/Bicep.LangServer/Deploy/DeploymentParametersHelper.cs

```diff
@@ -0,0 +1,181 @@
+  ""contentVersion"": ""1.0.0.0"",
+  ""parameters"": {
+  }
+}";
+                // We will send across the secure param values to the azure sdk that handles deployment,
+                // but will avoid writing it to the parameters file for security reasons
+                var updatedParametersFile = !string.IsNullOrWhiteSpace(parametersFilePath) ?
+                    File.ReadAllText(parametersFilePath) : armSchemaStyleParametersFile;
+                var updatedParametersFileWithoutSecureParams = updatedParametersFile;
+
+                var jObject = GetParametersObjectValue(updatedParametersFile, out bool isArmStyleTemplate);
+
+                foreach (var updatedDeploymentParameter in updatedDeploymentParameters.Reverse())
+                {
+                    var name = updatedDeploymentParameter.name;
+
+                    // Check to make sure parameters mentioned in parameters file are not overwritten
+                    if (jObject.ContainsKey(name))
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        var jsonEditor = new JsonEditor(updatedParametersFile);
+
+                        var propertyPaths = new List<string>();
+                        if (isArmStyleTemplate)
+                        {
+                            propertyPaths.Add("parameters");
+                            propertyPaths.Add(name);
+                        }
+                        else
+                        {
+                            propertyPaths.Add(name);
+                        }
+
+                        var valueObject = UpdateJObjectBasedOnParameterType(
+                            updatedDeploymentParameter.parameterType,
+                            updatedDeploymentParameter.value,
+                            JObject.Parse("{}"));
+
+                        (int line, int column, string text)? insertion = jsonEditor.InsertIfNotExist(propertyPaths.ToArray(), valueObject);
+
+                        if (insertion.HasValue)
+                        {
+                            var (line, column, insertText) = insertion.Value;
+
+                            updatedParametersFile = JsonEditor.ApplyInsertion(updatedParametersFile, (line, column, insertText));
+
+                            if (!updatedDeploymentParameter.isSecure)
+                            {
+                                updatedParametersFileWithoutSecureParams = JsonEditor.ApplyInsertion(updatedParametersFileWithoutSecureParams, (line, column, insertText));
+                            }
+                        }
+                    }
+                }
+
+                if (updatedDeploymentParameters.Any())
+                {
+                    if (updateOrCreateParametersFile == ParametersFileUpdateOption.Update)
+                    {
+                        File.WriteAllText(parametersFilePath, updatedParametersFileWithoutSecureParams);
```

Comment:
This could throw, e.g. read-only file.  Catch and throw better error than "invalid params file"?

Reply:
Actually I guess that's good ebnough.